### PR TITLE
Respect "Mask Passwords" Plugin

### PR DESCRIPTION
**Pull request by rwtsoftware**
Respect Mask Password plugin configuration
If there is a mask passwords wrapper associated with this build then:
* Grab password list from the wrapper
* Return `MaskPasswordsOutputStream` decorated with `LogstashOutputStream`

### CODE REVIEW

**Comment by R-Gerard** [comment]
So far so good. Looks like the next step is to access the build wrapper object and get the sensitive build variables.

I think you'll need to do something like this:

```
if (build.getProject() instanceof BuildableItemWithBuildWrappers) {
  BuildableItemWithBuildWrappers project = (BuildableItemWithBuildWrappers) build.getProject();
  for (BuildWrapper wrapper: project.getBuildWrappersList()) {
    if (wrapper instanceof MaskPasswordsBuildWrapper) {
      ...
    }
```

**Comment by rwtsoftware** [comment]
I have found an issue with this change as it is currently.

File: pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.573</version>
+    <version>1.599</version>
```
Comment (R-Gerard) [inline pom.xml]: What is the significance of this version of Jenkins?

File: pom.xml [hunk 2]
```diff
@@ -88,6 +88,18 @@
...
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>mask-passwords</artifactId>
+      <version>2.8</version>
+    </dependency>
+
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>junit</artifactId>
+      <version>1.10</version>
+    </dependency>
```
Comment (R-Gerard) [inline pom.xml]: Why are you depending on the junit plugin?

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 3]
```diff
@@ -67,7 +76,26 @@ public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener l
...
+        if (wrapper instanceof MaskPasswordsBuildWrapper) {
+          List<VarPasswordPair> allPasswordPairs = new ArrayList<VarPasswordPair>();
+
+          MaskPasswordsBuildWrapper maskPasswordsWrapper = (MaskPasswordsBuildWrapper) wrapper;
+          allPasswordPairs.addAll(maskPasswordsWrapper.getVarPasswordPairs());
+
+          MaskPasswordsConfig config = MaskPasswordsConfig.getInstance();
+          allPasswordPairs.addAll(config.getGlobalVarPasswordPairs());
+
+          return (OutputStream) los.maskPasswords(allPasswordPairs);
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: MaskPasswordsOutputStream is a descendant of OutputStream, so you don't need the cast here.

`return los.maskPasswords(allPasswordPairs);`

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 4]
```diff
@@ -67,7 +76,26 @@ public Environment setUp(AbstractBuild build, Launcher launcher, BuildListener l
...
+        }
+      }
+    }
+
+    return (OutputStream) los;
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: `return los;`

File: src/main/java/jenkins/plugins/logstash/LogstashOutputStream.java [hunk 5]
```diff
@@ -47,6 +52,15 @@ public LogstashOutputStream(OutputStream delegate, LogstashWriter logstash) {
     this.logstash = logstash;
   }
 
+  public MaskPasswordsOutputStream maskPasswords(List<VarPasswordPair> passwords) {
+    List<String> passwordStrings = new ArrayList<String>();
+    for (VarPasswordPair password: passwords)
+    {
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashOutputStream.java]: Minor stylistic nitpick: keep the curly brace with the previous line.

File: pom.xml (hunk 1 above)
Comment (rwtsoftware) [inline pom.xml]: This was the version that `SimpleBuildWrapper` was introduced.

File: pom.xml (hunk 2 above)
Comment (rwtsoftware) [inline pom.xml]: `hudson.tasks.test` was moved into the junit plugin somewhere between version `1.573` and `1.599`

File: pom.xml (hunk 1 above)
Comment (R-Gerard) [inline pom.xml]: You're not making use of `SimpleBuildWrapper` though.

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 6]
```diff
@@ -29,14 +29,23 @@
 import hudson.model.BuildListener;
 import hudson.model.AbstractBuild;
 import hudson.model.AbstractProject;
+import hudson.model.BuildableItemWithBuildWrappers;
 import hudson.tasks.BuildWrapper;
 import hudson.tasks.BuildWrapperDescriptor;
 
+import jenkins.tasks.SimpleBuildWrapper;
```
Comment (R-Gerard) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: This is an unused import.

File: pom.xml (hunk 1 above)
Comment (rwtsoftware) [inline pom.xml]: I am not directly using `SimpleBuildWrapper` so I can remove the unused import has been removed. `MaskPasswordsBuildWrapper` is a directly extends `SimpleBuildWrapper`. The build fails with build version `1.573`.

Comment (R-Gerard) [inline pom.xml]: Got it. I missed that on my first pass.

**Comment by Spitfire1900** [comment]
is it know if this works with the credentials-binding-plugin?

**Comment by Spitfire1900** [comment]
I've confirmed that this should work with credentials-binding-plugin as that plugin adds credentials as sensitive build variables and this prunes variables with the result of getSensitiveBuildVariables.

File: src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java [hunk 7]
```diff
@@ -29,14 +29,21 @@
...
 import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
 
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsBuildWrapper;
+import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsBuildWrapper.VarPasswordPair;
+import com.michelin.cio.hudson.plugins.maskpasswords.MaskPasswordsConfig;
+
```
Comment (Spitfire1900) [inline src/main/java/jenkins/plugins/logstash/LogstashBuildWrapper.java]: We have encountered issues with various plugins when using the mask password plugin. Could we use reflection to perform mask-password logic only if the mask password plugin is installed?

**Comment by rwtsoftware** [comment]
I am using the [credentials-binding-plugin](https://github.com/jenkinsci/credentials-binding-plugin) locally with this change. I use the [mask-passwords-plugin](https://github.com/jenkinsci/mask-passwords-plugin) to mask passwords that are injected through the [credentials-binding-plugin](https://github.com/jenkinsci/credentials-binding-plugin).

While I am un aware of an issue with the [mask-passwords-plugin](https://github.com/jenkinsci/mask-passwords-plugin) if there is one it should be fixed in that plugin. I don't recommend obscuring/complicating this implementation instead of fixing the root cause problem.

**Comment by Spitfire1900** [comment]
With the addition of these code changes would the Mask Password's plugin have to be installed as a dependency in Jenkins or are we just importing and using the jar at compile/run time?

File: pom.xml [hunk 8]
```diff
@@ -88,6 +88,18 @@
       <type>jar</type>
       <scope>compile</scope>
     </dependency>
+
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>mask-passwords</artifactId>
+      <version>2.8</version>
```
Comment (R-Gerard) [inline pom.xml]: Add `<optional>true</optional>` to the dependency.

**Comment by R-Gerard** [comment]
@Spitfire1900 As it is currently implemented, the Logstash plugin would require users to install the Mask Passwords plugin. Jenkins accomplishes this by looking at the build dependencies, and assumes all dependencies are required by default. In this case it's safe to make the dependency optional.

**Comment by Spitfire1900** [comment]
@rwtsoftware [The EnvInject plugin](https://wiki.jenkins-ci.org/display/JENKINS/EnvInject+Plugin) has incompatibilities with the Mask Passwords plugin. It can not retrieve information from it, so we wouldn't want to install both at once.

File: pom.xml (hunk 8 above)
Comment (rwtsoftware) [inline pom.xml]: Added

**Comment by Spitfire1900** [comment]
We have tested this on our own instance of Jenkins (with EnvInject plugin installed and Mask Passwords not installed) and it's working.

**Comment by R-Gerard** [comment]
Thanks again for the submission, @rwtsoftware. And thank you @Spitfire1900 for testing and feedback.

**Pull request merged**